File: tests/fixtures/sample-python-project/services/user_service/models/user_model.py
from sqlalchemy import Column, String, DateTime, Boolean, Integer, ForeignKey, func
from sqlalchemy.orm import relationship, DeclarativeBase


class Base(DeclarativeBase):
    pass


class User(Base):
    __tablename__ = "users"

    id = Column(String, primary_key=True)
    email = Column(String, unique=True, nullable=False)
    name = Column(String, nullable=True)
    posts = relationship("Post", back_populates="author")
    created_at = Column(DateTime, server_default=func.now())
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now())


class Post(Base):
    __tablename__ = "posts"

    id = Column(String, primary_key=True)
    title = Column(String, nullable=False)
    content = Column(String, nullable=True)
    published = Column(Boolean, default=False)
    author_id = Column(String, ForeignKey("users.id"), nullable=False)
    author = relationship("User", back_populates="posts")
    created_at = Column(DateTime, server_default=func.now())


# Intentional violations for analysis rule testing:
# - Missing timestamps (no created_at/updated_at)
# - category_id has no FK relation (missing foreign key)
# - Overly nullable columns
# - Naming inconsistency: mixes snake_case (category_id) with camelCase (viewCount)
class Comment(Base):
    __tablename__ = "comments"

    id = Column(String, primary_key=True)
    body = Column(String, nullable=True)
    category_id = Column(String, nullable=True)
    post_id = Column(String, nullable=True)
    rating = Column(Integer, nullable=True)
    viewCount = Column(Integer, nullable=True)
    status = Column(String, default="draft")


# Another model with issues:
# - Missing timestamps
# - tag_id column with no FK relation
class PostTag(Base):
    __tablename__ = "post_tags"

    id = Column(String, primary_key=True)
    tag_id = Column(String, nullable=False)
    label = Column(String, nullable=True)

File: tests/fixtures/sample-csharp-project-negative/services/UserService/Models/UserModel.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using Microsoft.EntityFrameworkCore;

namespace UserServiceApp.Models;

public class AppDbContext : DbContext
{
    public DbSet<User> Users { get; set; } = null!;
    public DbSet<Post> Posts { get; set; } = null!;
    public DbSet<Comment> Comments { get; set; } = null!;
    public DbSet<PostTag> PostTags { get; set; } = null!;

    protected override void OnConfiguring(DbContextOptionsBuilder optionsBuilder)
    {
        optionsBuilder.UseNpgsql("Host=localhost;Port=5432;Database=app;Username=app;Password=secret");
    }
}

public class User
{
    [Key]
    public string Id { get; set; } = null!;

    [Required]
    public string Email { get; set; } = null!;

    public string? Name { get; set; }

    public List<Post> Posts { get; set; } = new();

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;

    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public class Post
{
    [Key]
    public string Id { get; set; } = null!;

    [Required]
    public string Title { get; set; } = null!;

    public string? Content { get; set; }

    public bool Published { get; set; } = false;

    [ForeignKey("User")]
    public string UserId { get; set; } = null!;

    public User User { get; set; } = null!;

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
}

public class Comment
{
    [Key]
    public string Id { get; set; } = null!;

    public string? Body { get; set; }

    public string? CategoryId { get; set; }

    public string? PostId { get; set; }

    public int? Rating { get; set; }

    public int? ViewCount { get; set; }

    public string Status { get; set; } = "draft";
}

public class PostTag
{
    [Key]
    public string PostId { get; set; } = null!;

    public string? Tag { get; set; }

    public string? Label { get; set; }
}
